libpod/pod.go

```diff
@@ -71,19 +72,167 @@ func (p *Pod) Init() error {
+			return err
+		}
+
+		logrus.Debugf("Started container %s", ctr.ID())
+	}
+
+	return nil
 }
 
 // Stop stops all containers within a pod that are not already stopped
+// Each container will use its own stop timeout
+// Only running containers will be stopped. Paused, stopped, or created
+// containers will be ignored.
+// If an error is encountered stopping any one container, no further containers
+// will be stopped, and an error will immediately be returned.
+// Stop() is not an atomic operation - if an error is encountered, containers
+// which have already been stopped will not be restarted
 func (p *Pod) Stop() error {
-	return ErrNotImplemented
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if !p.valid {
+		return ErrPodRemoved
+	}
+
+	allCtrs, err := p.runtime.state.PodContainers(p)
+	if err != nil {
+		return err
+	}
+
+	// We need to lock all the containers
+	for _, ctr := range allCtrs {
+		ctr.lock.Lock()
+		defer ctr.lock.Unlock()
+
+		if err := ctr.syncContainer(); err != nil {
+			return err
+		}
+	}
+
+	// Send a signal to all containers
+	for _, ctr := range allCtrs {
+		// Ignore containers that are not running
+		if ctr.state.State != ContainerStateRunning {
+			continue
+		}
+
+		if err := ctr.runtime.ociRuntime.stopContainer(ctr, ctr.config.StopTimeout); err != nil {
+			return errors.Wrapf(err, "error stopping container %s", ctr.ID())
+		}
+
+		// Sync container state to pick up return code
+		if err := ctr.runtime.ociRuntime.updateContainerStatus(ctr); err != nil {
+			return err
+		}
+
+		// Clean up storage to ensure we don't leave dangling mounts
+		if err := ctr.cleanupStorage(); err != nil {
+			return err
+		}
+
+		logrus.Debugf("Stopped container %s", ctr.ID())
+	}
+
+	return nil
 }
 
 // Kill sends a signal to all running containers within a pod
+// Signals will only be sent to running containers. Containers that are not
+// running will be ignored.
+// If an error is encountered signalling any one container, kill will stop
+// and immediately return an error, sending no further signals
+// Kill() is not an atomic operation - if an error is encountered, no further
+// signals will be sent, but some signals may already have been sent
 func (p *Pod) Kill(signal uint) error {
-	return ErrNotImplemented
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if !p.valid {
+		return ErrPodRemoved
+	}
+
+	allCtrs, err := p.runtime.state.PodContainers(p)
+	if err != nil {
+		return err
+	}
+
+	// We need to lock all the containers
+	for _, ctr := range allCtrs {
+		ctr.lock.Lock()
+		defer ctr.lock.Unlock()
+
+		if err := ctr.syncContainer(); err != nil {
+			return err
+		}
+	}
+
+	// Send a signal to all containers
+	for _, ctr := range allCtrs {
+		// Ignore containers that are not running
+		if ctr.state.State != ContainerStateRunning {
+			continue
+		}
+
+		if err := ctr.runtime.ociRuntime.killContainer(ctr, signal); err != nil {
+			return errors.Wrapf(err, "error killing container %s", ctr.ID())
+		}
+
```

Comment:
Ditto ctr.save() here?

Reply:
We don't know for sure the container's status changed, so we let this go without changing or saving anything. The next command should do a syncContainer() which will pick up the changes (if any) and save them to the DB

Reply:
Is there any chance of a race condition here?  Where the container was killed before you send the kill signal, and the killContainer will fail because the container is not running.

I am not a big fan of us not being able to kill all containers or stop all containers because one of the stop or kill signal fails.  This could leave the user in a state where it is difficult to kill/stop a pod if the first container is in a weird state.

Reply:
Since the container is locked, we can't manually stop or start it. There is the case where it dies naturally on its own between us checking the status and sending the signal, but that shouldn't cause any issues with the state. The container will fail to send the signal with an error and the next command on it will pick up that it stopped.

Reply:
But will ctr.runtime.ociRuntime.killContainer() return and error causing the rest of the kills to fail?

Reply:
Yeah, it should. If we make all the kills independent that would resolve the issue